[WIP] Fix struct/const revision

timholy opened the pull request:
The aim here is to provide full support for struct/const revision. This first commit just adds simple tests, but it already reveals something that needs to be addressed: when a `struct` gets revised, methods defined for that `struct` aren't automatically re-evaluated for the new type. Specifically, in the added tests, after revising the definition of `Point` we get
```julia
julia> methods(StructConst.mynorm)
# 1 method for generic function "mynorm" from StructConst:
 [1] mynorm(p::@world(StructConst.Point, 40598:40738))
     @ /tmp/jW7BpD8RL7/StructConst/src/StructConst.jl:12
```
and there is no defined method for a currently-valid `StructConst.Point`.

@Keno, if you have a moment, let me check my understanding of the current situation:

- method applicability is assessed by type-intersection, as usual, but now types can have a world-age
- the MethodInstances and CodeInstances involving the old type are still valid (their world age is not capped, but of course they require inputs with "old" types)
- there is no invalidation step needed for changes of the kind tested with the revision of Point (`invalidate_code_for_globalref!` is not called?)
- we can't count on `binding.backedges` to list everything that uses this binding
- the right solution is for Revise to:
  + detect that we're about to redefine a struct
  + search the entire system for any methods where that type appears in a signature
  + parse the source files, if necessary
  + redefine the struct
  + re-evaluate the method definitions

Presumably, waiting to do the last step as the last stage of a revision would be wise, as it is possible that more than one struct will be revised at the same time, and one might as well do each evaluation only once.

Keno commented:
> * method applicability is assessed by type-intersection, as usual, but now types (via their bindings) can have a world-age

I think that's a valid way to say it, but just to clarify, nothing about the type object itself changes, just the name by which you have to access it

> * the MethodInstances and CodeInstances involving the old type are still valid (their world age is not capped, but of course they require inputs with old types)

yes

> * there is no invalidation step needed for changes of the kind tested with the revision of Point (`invalidate_code_for_globalref!` is not called?)

I don't understand what you're asking

> * we can't count on `binding.backedges` to list everything that uses this binding

It lists all cross-module edges to get everything, you also need to iterate all methods in the same module as the binding.

> * the right solution is for Revise to:
>   
>   * detect that we're about to redefine a struct
>   * search the entire system for any methods where that type appears in a signature
>   * parse the source files, if necessary
>   * redefine the struct
>   * re-evaluate the method definitions

In the fullness of time, the correct way to implement Revise is to run the toplevel expressions in a non-standard evaluation mode that tracks dependency edges for any evaluated constants. However, that should maybe wait until after the JuliaLowering changes and the interpreter rewrite. I think your proposed strategy is reasonable in the meantime although not fully complete.

timholy commented:
> > there is no invalidation step needed for changes of the kind tested with the revision of Point (`invalidate_code_for_globalref!` is not called?)
> 
> I don't understand what you're asking

I was wondering whether some kind of callback in there could be helpful in short-circuiting the process of determining which methoddefs need to be re-evaluated, but if it's not called then that won't help. From what I can tell, that's not going to work anyway because it only traverses the module containing the type definition. But your comment

> It lists all cross-module edges to get everything, you also need to iterate all methods in the same module as the binding.

seems to be what I was looking for.

This seems fairly straightforward, thanks for the excellent groundwork.

Keno commented:
> I was wondering whether some kind of callback in there could be helpful in short-circuiting the process of determining which methoddefs need to be re-evaluated

No, there's no edges of any kind for signatures. You need to scan the entire system. However, an earlier version of that code did the same whole-system scan, so you can steal some code for whole-system scanning.

Keno commented:
There's two kinds of edges that are tracked there. One is explicit import/using. The other is to lowered code of method definitions (but only after the first inference). At no point is an edge ever added for an evaluation of a binding, only for compilation of code that references the binding.

timholy commented:
Just to cross-check:
```
                module StructConstUser
                using StructConst: Fixed, Point
                struct PointWrapper
                    p::Point
                end
                scuf(f::Fixed) = 33 * f.x
                scup(p::Point) = 44 * p.x
                end
```
yields a backedge for the `using StructConst: Fixed, Point` statement, but I don't see anything related to `scup` even if I evaluate it:
```julia
julia> e = only(b.backedges)
Binding StructConstUser.Point
   40598:∞ - explicit `using` from StructConst.Point
   1:0 - undefined binding - guard entry
```

Keno commented:
> but I don't see anything related to `scup` even if I evaluate it:

That's because `scup` doesn't use any bindings in its lowered code (as I said, no edges for signatures). However, even if you had like `mkscup() = Point()`, that binding would be to StructConstUser. To actually get a cross-module edge is not that easy, but you could do `@eval mkscup() = $(GlobalRef(StructConst, :Point))()`. The most common case to get cross-module edges is bindings. Also, not a lot of edges is good, that is the design goal, since we don't want to store them :).

timholy commented:
This is getting close, but there's a philosophical question to answer: should Revise try to hew to Base behavior as closely as possible, or should it add/modify behaviors for what might be a nicer interactive experience?

In this case, the issue is the following: when I redefine `Point`, in Base you can still call previously-defined functions on "old data." However, in this PR, calling functions on old data throws (or is intended to throw) a `MethodError`. Why might you want that? Because I imagine that a common user-error will be to forget to refresh your data with updated types, and if your whole code pipeline runs (using all the old definitions), I imagine people will bang their heads against walls trying to figure out why their edits aren't having the intended effect. In contrast, if they get a `no method foo(::@world(Point, m:n))`, they'll pretty quickly learn how you fix it.

I should say that initially I wasn't aiming in this direction, and this current proposal arose from noticing some unexpected subtleties about the order in which Revise's cache files get set up: the order in which you parse the source, lower the code, and rebind the name matters quite a lot, and some of the choices in this version are designed to compensate for the fact that Revise normally doesn't parse a file unless it's been edited. (Now, we'll also have to parse any file with a method whose signature references an updated type.) I think it's possible to hew to the Base behavior, if we decide that's better, but the initial bugs I faced led me to ask what behavior we actually want, and I came to the conclusion that it's likely a better user experience if we invalidate methods that only work on old types.

timholy commented on src/packagedef.jl:
```diff
@@ -874,6 +882,33 @@ function revise(; throw=false)
+                end
+                Base.delete_method(m)  # ensure that "old data" doesn't get run with "old methods"
+                _, ex = methinfo[1]
+                eval_with_signatures(m.module, ex; mode=:eval)
```
Note to self: I was surprised that dependent types (e.g., `PointWrapper`) don't experience binding invalidation until their constructors are re-evaluated. `Core.eval(m.module, ex)` triggers the binding invalidation and gets us farther in the tests, but it doesn't add the appropriate signatures to `CodeTracking.method_info`. My sense is that JuliaInterpreter/JuliaLowering need to trigger binding invalidation from `LoweredCodeUtils.methoddef!`.

Keno commented:
> However, in this PR, calling functions on old data throws (or is intended to throw) a `MethodError`. Why might you want that?

I think that's fine. Conceptually Revise does both adding new methods and deleting old ones. Base is always "append only".

timholy commented:
The collateral damage to tests is something I'm looking into; if one runs all the tests, there appear to be some failures that predate this effort, including an incomplete updating of the ecosystem to https://github.com/JuliaLang/julia/pull/52415. I'll work on that.

But meanwhile, this seems to be working:

```julia
tim@diva:~/.julia/dev/Revise$ ~/src/juliaw/julia --startup=no --project -e 'using Pkg; Pkg.test(; test_args=["struct/const revision"], coverage=false)'
     Testing Revise
      Status `/tmp/jl_4EbO8B/Project.toml`
# Pkg output deleted
     Testing Running tests...
Test Summary: | Pass  Total   Time
Revise        |   30     30  10.3s
beginning cleanup
```

and may be ready for anyone who wants to review it.

aviatesk commented:
This is interesting.

So it looks like this PR tracks edges not only from binding to method signature, but also from binding to method body?

I think a similar mechanism will be needed when developing a new language server. If it's already implemented in Revise, I'd like to reuse it.

I haven't had a chance to look at this PR closely yet, but I'll read it later.

timholy commented:
I'll be curious to see whether it improves the efficiency of hacking on inference.

Keno commented:
> There's a separate bit of novelty for `Core.Compiler`: this PR also checks whether any of the entry points to type inference have been invalidated, and if so calls `bootstrap!`

I don't know if I like that. We already have `@activate Compiler[:codegen]` for this use case.

timholy commented:
I didn't know about `@activate`. Is there a demo of the workflow somewhere?

I can back that commit out, if there are other mechanisms to achieve the same aim.

KristofferC commented:
I tired this. On:

- Revise (this PR)
- LoweredCodeUtils master
- Julia 1.12.0-beta1

What I did was:

- Load Revise, Tensors.
- Make the following revision to Tensors: 
  ```diff
  diff --git a/src/Tensors.jl b/src/Tensors.jl
  index 4cd677b..5f04312 100644
  --- a/src/Tensors.jl
  +++ b/src/Tensors.jl
  @@ -63,6 +63,7 @@ julia> Tensor{1,3,Float64}((1.0, 2.0, 3.0))
   """
   struct Tensor{order, dim, T, M} <: AbstractTensor{order, dim, T}
       data::NTuple{M, T}
  +    x::Int
       Tensor{order, dim, T, M}(data::NTuple) where {order, dim, T, M} = new{order, dim, T, M}(data)
   end
  ```
- Press 1 in the REPL

with the following result:

```julia
julia> using Revise, Tensors

# updates Tensors file here

julia> 1
ERROR: UndefVarError: `Broadcasted` not defined in `StaticArrays.StableFlatten`
Suggestion: check for spelling errors or missing imports.
Stacktrace:
 [1] top-level scope
   @ ~/.julia/packages/StaticArrays/LSPcF/src/broadcast.jl:180
Revise evaluation error at /Users/kristoffercarlsson/.julia/packages/StaticArrays/LSPcF/src/broadcast.jl:180

Stacktrace:
  [1] methods_by_execution!(recurse::Any, methodinfo::Revise.CodeTrackingMethodInfo, docexprs::Dict{…}, mod::Module, ex::Expr; mode::Symbol, disablebp::Bool, always_rethrow::Bool, kwargs::@Kwargs{})
    @ Revise ~/JuliaPkgs/Revise.jl/src/lowered.jl:306
  [2] eval_with_signatures(mod::Module, ex::Expr; mode::Symbol, kwargs::@Kwargs{})
    @ Revise ~/JuliaPkgs/Revise.jl/src/packagedef.jl:556
  [3] eval_with_signatures
    @ ~/JuliaPkgs/Revise.jl/src/packagedef.jl:553 [inlined]
  [4] instantiate_sigs!(modexsigs::OrderedCollections.OrderedDict{…}; mode::Symbol, kwargs::@Kwargs{})
    @ Revise ~/JuliaPkgs/Revise.jl/src/packagedef.jl:564
  [5] instantiate_sigs!
    @ ~/JuliaPkgs/Revise.jl/src/packagedef.jl:560 [inlined]
  [6] maybe_extract_sigs_for_meths(meths::Set{Method})
    @ Revise ~/JuliaPkgs/Revise.jl/src/pkgs.jl:162
  [7] (::Revise.var"#107#108"{Bool})()
    @ Revise ~/JuliaPkgs/Revise.jl/src/packagedef.jl:941
  [8] lock(f::Revise.var"#107#108"{Bool}, l::ReentrantLock)
    @ Base ./lock.jl:335
  [9] #revise#104
    @ ~/JuliaPkgs/Revise.jl/src/packagedef.jl:844 [inlined]
 [10] revise()
    @ Revise ~/JuliaPkgs/Revise.jl/src/packagedef.jl:842
 [11] top-level scope
    @ REPL:1

caused by: UndefVarError: `Broadcasted` not defined in `StaticArrays.StableFlatten`
Suggestion: check for spelling errors or missing imports.
Stacktrace:
  [1] lookup_var
    @ ~/.julia/packages/JuliaInterpreter/J7J9G/src/interpret.jl:5 [inlined]
  [2] step_expr!(recurse::Any, frame::JuliaInterpreter.Frame, node::Any, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/J7J9G/src/interpret.jl:44
  [3] signature(recurse::Any, frame::JuliaInterpreter.Frame, stmt::Any, pc::Int64)
    @ LoweredCodeUtils ~/.julia/packages/LoweredCodeUtils/BIVzf/src/signatures.jl:50
  [4] methoddef!(recurse::Any, signatures::Vector{Any}, frame::JuliaInterpreter.Frame, stmt::Any, pc::Int64; define::Bool)
    @ LoweredCodeUtils ~/.julia/packages/LoweredCodeUtils/BIVzf/src/signatures.jl:595
  [5] methoddef!
    @ ~/.julia/packages/LoweredCodeUtils/BIVzf/src/signatures.jl:534 [inlined]
  [6] methods_by_execution!(recurse::Any, methodinfo::Revise.CodeTrackingMethodInfo, docexprs::Dict{…}, frame::JuliaInterpreter.Frame, isrequired::Vector{…}; mode::Symbol, skip_include::Bool)
    @ Revise ~/JuliaPkgs/Revise.jl/src/lowered.jl:354
  [7] methods_by_execution!
    @ ~/JuliaPkgs/Revise.jl/src/lowered.jl:317 [inlined]
  [8] methods_by_execution!(recurse::Any, methodinfo::Revise.CodeTrackingMethodInfo, docexprs::Dict{…}, mod::Module, ex::Expr; mode::Symbol, disablebp::Bool, always_rethrow::Bool, kwargs::@Kwargs{})
    @ Revise ~/JuliaPkgs/Revise.jl/src/lowered.jl:296
  [9] eval_with_signatures(mod::Module, ex::Expr; mode::Symbol, kwargs::@Kwargs{})
    @ Revise ~/JuliaPkgs/Revise.jl/src/packagedef.jl:556
 [10] eval_with_signatures
    @ ~/JuliaPkgs/Revise.jl/src/packagedef.jl:553 [inlined]
 [11] instantiate_sigs!(modexsigs::OrderedCollections.OrderedDict{…}; mode::Symbol, kwargs::@Kwargs{})
    @ Revise ~/JuliaPkgs/Revise.jl/src/packagedef.jl:564
 [12] instantiate_sigs!
    @ ~/JuliaPkgs/Revise.jl/src/packagedef.jl:560 [inlined]
 [13] maybe_extract_sigs_for_meths(meths::Set{Method})
    @ Revise ~/JuliaPkgs/Revise.jl/src/pkgs.jl:162
 [14] (::Revise.var"#107#108"{Bool})()
    @ Revise ~/JuliaPkgs/Revise.jl/src/packagedef.jl:941
 [15] lock(f::Revise.var"#107#108"{Bool}, l::ReentrantLock)
    @ Base ./lock.jl:335
 [16] #revise#104
    @ ~/JuliaPkgs/Revise.jl/src/packagedef.jl:844 [inlined]
 [17] revise()
    @ Revise ~/JuliaPkgs/Revise.jl/src/packagedef.jl:842
 [18] top-level scope
    @ REPL:1
Some type information was truncated. Use `show(err)` to see complete types.
```

aviatesk commented:
I updated against the latest master branch.

I tested this branch with the simple case, but it seems that for `export`ed bindings, additional edge tracking is needed? 
In particular, the world ages of bindings that are `using`ed need to be updated, it seems.
For example, let's say we have an Example package like this:
```julia
module Example
export hello, Hello

struct Hello
    who::String
end

hello(x::Hello) = hello(x.who)

"""
    hello(who::String)

Return "Hello, `who`".
"""
hello(who::String) = "Hello, $who"

end
```
We get the following error:
```julia-repl
julia> using Revise

julia> using Example

julia> hello(Hello("world"))
"Hello, world"

julia> # Apply the following diff
       # ```diff
       # diff --git a/src/Example.jl b/src/Example.jl
       # index 65c7eae..c631814 100644
       # --- a/src/Example.jl
       # +++ b/src/Example.jl
       # @@ -2,10 +2,10 @@ module Example
       #  export hello, Hello
       #  
       #  struct Hello
       # -    who::String
       # +    who2::String
       #  end
       #  
       # -hello(x::Hello) = hello(x.who)
       # +hello(x::Hello) = hello(x.who2 * " (changed)")
       #  
       #  """
       #      hello(who::String)
       # ```

julia> hello(Hello("world"))
ERROR: MethodError: no method matching @world(Example.Hello, 38355:38383)(::String)
The type `@world(Example.Hello, 38355:38383)` exists, but no method is defined for this combination of argument types when trying to construct it.
Stacktrace:
 [1] top-level scope
   @ REPL[5]:1

julia> hello(Example.Hello("world")) # but this works
"Hello, world (changed)"
```

timholy commented:
@KristofferC your example looks like a selective evaluation issue:
https://github.com/JuliaArrays/StaticArrays.jl/blob/e7c5e0162fe829f317b1d49aea7f192eb8c58eae/src/broadcast.jl#L171-L175
doesn't run because of the Julia version number, but somehow Revise still tries to step through it.

aviatesk commented:
Thanks for fixing my example!

> The `return` statement is not marked

Sure, I will look into it at least by tomorrow.

aviatesk commented:
> I didn't know about `@activate`. Is there a demo of the workflow somewhere? I've seen the docs, and they're good, but I suspect there's a bit more wisdom around effective usage than they convey.
> 
> I can back that commit out, if there are other mechanisms to achieve the same aim. Are you saying that `Revise.track(Core.Compiler)` should be deprecated?

Sorry for the delay, but I'd like to add some details about this.

Currently, when you use `InteractiveUtils.@activate Compiler`, the Compiler.jl implementation specified in your environment gets loaded. This also switches the Compiler.jl implementation used by reflection utilities like `code_typed` to the newly loaded one.
For the actual Compiler.jl implementation, besides a dummy implementation that just redirects to `Base.Compiler` (https://github.com/JuliaLang/BaseCompiler.jl), we expect the implementation to be used is a local checkout of `JuliaLang/julia/Compiler` or a specific remote `JuliaLang/julia` commit.

So when you're hacking on Compiler.jl, using `InteractiveUtils.@activate Compiler` is generally enough, and I think `Revise.track(Core.Compiler)` can be deprecated. In most cases, I don't see any benefit in applying changes to the `Base.Compiler` module and observing the changes for the Compiler development. There might still be situations where it still is potentially useful, like for debugging inference issues that only happen during bootstrap. But this problem itself is quite tricky, and I feel it's better to use simple reflections like `Core.eval` or `Core.println` for debugging it, rather than advanced tools like Revise. So in conclusion, I think `Revise.track(Core.Compiler)` can be deprecated.

@Keno Please let me know if you have anything to add.

aviatesk commented:
For some reason, @KristofferC's example seems to be working fine with the latest combination of Revise/LoweredCodeUtils/Tensors.

```julia-repl
julia> using Revise

julia> using Tensors

julia> fieldnames(Tensors.Tensor)
(:data,)

julia> fieldnames(Tensors.Tensor)
(:data, :x)

julia> VERSION
v"1.12.0-rc1.23"
```

timholy commented:
Not for me (or at least not on nightly).

timholy commented:
Amusingly/terrifyingly, I have even occasionally triggered the message `"Compiling the compiler. This may take several minutes ..."` just from using the REPL and making revisions 😱

timholy commented:
Well, that was easy. This branch may now be worth trying more seriously.

Keno commented:
I thought I had commented, but I guess not

> I think `Revise.track(Core.Compiler)` can be deprecated

This is very occasionally still useful, when the `Compiler` in your environment is not the same as the base compiler and you want to test things. So I don't think we need to remove it, we can just teach people about the new, better workflow for Compiler development, and `Revise.track` can remain an option for power users.

timholy commented:
struct revision tests are now passing again (the failure is only in `extra tests`).

In principle, this could be ready to merge. Please test in real workflows.

MasonProtter commented:
> Please test in real workflows (nightly, as rc1 is currently broken).

Can you suggest a branch that does work?

timholy commented:
`nightly`

MasonProtter commented:
My bad, I completely misread the comment as saying that both `nightly` and `1.12` were broken. Sorry for the noise.

timholy commented:
Oh weird:
```
tim@hypnotic:/tmp/rvs$ julia +nightly -q
julia> includet("foobar.jl")
WARNING: Detected access to binding `Main.Foo` in a world prior to its definition world.
  Julia 1.12 has introduced more strict world age semantics for global bindings.
  !!! This code may malfunction under Revise.
  !!! This code will error in future versions of Julia.
Hint: Add an appropriate `invokelatest` around the access to this binding.
To make this warning an error, and hence obtain a stack trace, use `julia --depwarn=error`.

julia>
tim@hypnotic:/tmp/rvs$ julia +nightly -q --depwarn=error
julia> includet("foobar.jl")

julia>
```

?!?

timholy commented:
@lassepe, this behavior observed at the end of your demo is quite surprising:

```julia
julia> fieldnames(Bar)
()

julia> fieldnames(Foo.Bar)
(:x,)

julia> Foo.Bar(1)
Main.Foo.Bar(1)
```

@Keno, is this a Julia bug failing to propagate the updated binding to `Main`?

timholy commented:
https://github.com/JuliaLang/julia/issues/59272

jakobnissen commented:
I can't get Revise to work at all with this PR. MWE:
* Make a new package `Foo` with the content (in `src/Foo.jl`):
```julia
module Foo

struct Foo
    x::Int
end

foo() = 2

end # module Foo
```
* Launch Julia 1.12.0-rc2, with an environment only containing `Foo` and Revise with this PR (`https://github.com/timholy/Revise.jl.git#teh/struct_revision`), no startup file
* In the REPL, type `julia> using Revise, Foo; Foo.foo()`. It works correctly
* Change `foo() = 2` to `foo() = 1`
* Now, in the REPL:
```julia
julia> Foo.foo()
┌ Error: Failed to revise /tmp/Foo/src/Foo.jl
│   exception =
│    invalid redefinition of constant Foo
│    Stacktrace:
│     [1] queuenext!(iter::JuliaInterpreter.ExprSplitter)
│       @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/378J1/src/construct.jl:492
└ @ Revise ~/.julia/packages/Revise/FV0Wc/src/packagedef.jl:785
┌ Warning: The running code does not match the saved version for the following files:
│
│   /tmp/Foo/src/Foo.jl
│
│ If the error was due to evaluation order, it can sometimes be resolved by calling `Revise.retry()`.
│ Use Revise.errors() to report errors again. Only the first error in each file is shown.
│ Your prompt color may be yellow until the errors are resolved.
└ @ Revise ~/.julia/packages/Revise/FV0Wc/src/packagedef.jl:957
2
```

lassepe commented:
@jakobnissen, isn't the problem simply that both your module and the struct are called `Foo`? I think that never worked.

lassepe commented:
A modified version of my original setting above still seems to be broken (in a different way) on julia 1.12-rc2 and the lateset version of this branch (


**Steps to reproduce**

1. Generate a package: `]generate StructRevision`
2. Add the following content to `src/StructRevision`
```julia
module StructRevision

struct Foo
  x::Int
end

bar(::Foo) = 1

end # module StructRevision
```
3. With julia 1.12-rc2 from the REPL:
```julia
❯ julia +1.12 --startup-file=no --project
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.12.0-rc2 (2025-09-06)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org release
|__/                   |

julia> using Revise, StructRevision

julia> foo1 = StructRevision.Foo(1)
StructRevision.Foo(1)

julia> StructRevision.bar(foo1)
1

julia> # after changing Foo.x from Int to Float64:

julia> foo2 = StructRevision.Foo(1)
StructRevision.Foo(1.0)

julia> StructRevision.bar(foo1)
1

julia> StructRevision.bar(foo2)
ERROR: MethodError: no method matching bar(::StructRevision.Foo)
The function `bar` exists, but no method is defined for this combination of argument types.

Closest candidates are:
  bar(::@world(StructRevision.Foo, 38519:38547))
   @ StructRevision ~/worktree/BugReports/StructRevision.jl/src/StructRevision.jl:7

Stacktrace:
 [1] top-level scope
   @ REPL[7]:1
```

timholy commented:
> Or are there known serious regressions from this?

Yes (e.g., https://github.com/timholy/Revise.jl/pull/894#issuecomment-3279854151). When it fails, it's strictly worse than just using `master` (and not having good struct revision).

aviatesk commented:
Ah, it's been a long journey. This work is finally wrapping up.

I think Revise can now handle many cases for updating `struct` definitions. On my end, I've confirmed that all the issues brought up in this long thread are resolved, so it would be great if you could give it another try when you have some time.

Starting today, I'll spend a few days testing it, and the plan is to merge it around the new year.

@timholy I've re-evaluated and implemented the invalidations Revise needs to handle to enable struct revision in most of realistic cases. This ended up adding nearly 1000 new lines of changes, and the logic has been significantly rewritten from the initial struct revision implementation, so I'd like to leave a few points to note for future reference:
- To add information to the `pkgdatas` cache about which type definition an expression is associated with, I've generalized the `ModuleExprsSigs` type to `ModuleExprsInfos`, and its leaf type has been changed from the previous `SigInfo` to `Union{SigInfo,TypeInfo}`.
    - However, for `TypeInfo`, I haven't prepared a lightweight cache on the CodeTracking side that corresponds to `CodeTracking.method_info` yet. Revise didn't need it for a correct implementation, and it would be a hassle to maintain compatibility with CodeTracking, so I decided it's not worth doing right now. But it might be useful for optimization, so I've left a TODO comment in `packagedef.jl` just in case.
- Thanks to Julia's binding partition support, Revise could, in principle, correctly propagate the effects associated with binding redefinition to the active session (regardless of whether the binding is `const` or not), not just the struct revision implemented in this PR, by cleverly using `delete_binding` and similar functions. This is a superset of the "struct revision" implemented by this PR, which I'm calling "binding revision.", meaning it enables revision in cases like the following:
  ```julia
    global n::Int = 10 # and then modified to `n::Int = 5`
    for i = 1:n
        @eval func(::Val{$i}) = $i
    end
    ```
    However, to handle this correctly, we'd need to explicitly manage access to bindings accompanied with top-level code execution and track their edges, which is quite technically challenging. I might do that in the future, but I don't know when. For now, this PR only achieves "struct revision," so I've updated the PR description based on that.
- Regarding the invalidation logic that happens with type redefinition, you can understand it by looking at `record_invalidations_for_type_deletion!`. Tests were added also.
- I've removed the logic that automatically performs Compiler.jl's bootstrap. I wasn't entirely confident if the logic was correct, and I also prefer calling `bootstrap` manually. Also, I didn't think there was a strong need to add such additional revision logic specifically for Compiler.jl, so I've removed it for now.

- One of the minor things I may finish before merging is the revision ordering problem. Currently, revision follows a simple flow like this:
    - First, execute the code modified by the user.
    - Redefine type definitions related to that code.
    - Redefine methods related to that code.
  However, this approach might cause errors if the user's modified code includes methods that should be defined using new types. We might need to implement a revision order that first defines all related type definitions, including those changed by the user, and then redefines related method definitions. But for now, the current logic works fine in the scope of my testing, so I'm keeping this naive logic for the time being.

aviatesk commented:
@jkrumbiegel As far as I've tested on my end, the issue you found seems to be resolved. It would be helpful if you could try it again on your side too.
```julia-repl
julia> using Revise
Precompiling Revise finished.
  1 dependency successfully precompiled in 6 seconds

julia> using CairoMakie

julia> fieldnames(CairoMakie.Makie.ScrollZoom)
(:speed, :reset_timer, :prev_xticklabelspace, :prev_yticklabelspace, :reset_delay)

julia> lines(cumsum(randn(100)))

julia> fieldnames(CairoMakie.Makie.ScrollZoom) # after edit (reset_timer -> reset_time)
(:speed, :reset_time, :prev_xticklabelspace, :prev_yticklabelspace, :reset_delay)

julia> lines(cumsum(randn(100)))
```

aviatesk commented:
I ran the same test on Makie and CairoMakie with https://github.com/MakieOrg/Makie.jl/pull/5402 reverted, and it seems to be working without warnings.

topolarity commented:
Same limitation seems to apply for type-aliases:
```julia
IdArray{N} = Array{Int64,N} # change and `Graph` is not updated
struct Graph
    allocated::IdArray{1}
    conflicted::IdArray{1}
    resolved::IdArray{1}
end
```
Is that expected?

This one is a little weird, because it _is_ a type you're re-defining at top-level. It's just not a `struct` definition.

aviatesk commented:
> Same limitation seems to apply for type-aliases:
> 
> ```julia
> IdArray{N} = Array{Int64,N} # change and `Graph` is not updated
> struct Graph
>     allocated::IdArray{1}
>     conflicted::IdArray{1}
>     resolved::IdArray{1}
> end
> ```
> 
> Is that expected?
> 
> This one is a little weird, because it _is_ a type you're re-defining at top-level. It's just not a `struct` definition.

Thanks for testing this! And yeah, Revise doesn't handle this case right now. It's what I'd call a 'binding revision' case, rather than a 'struct revision'. But I think we could support it by adding a simple 'ad-hoc hack'. The same goes for the first case you mentioned; we could add that before merging this PR (added a new TODO item for the OP).
Generalizing the hack would require some pretty advanced techniques (we have ideas on how to handle it, but I won't go into the specific methods here for various reasons).

aviatesk commented:
@lassepe, thanks for testing it. Glad it worked for your case.

aviatesk commented:
Alright, let's merge this.
After merging, I'll update NEWS.md and release v3.13.
